### PR TITLE
Add Location class

### PR DESCRIPTION
**Pull request by iwootten**
### What is the context of this PR?

Updates survey runner to use a Location class for representation of group/instance/block locations and updates navigator to return lists of them when requesting paths. 

Allows us to generate urls from locations when metadata is known, which negates the need for jinja filter to create navigation links.

Updated version of https://github.com/ONSdigital/eq-survey-runner/pull/661, based on huge amount of work gone in of late.

### How to review 
Take a look at the approach, see if you approve.

### CODE REVIEW

File: app/main/views/questionnaire.py
Comment (collisdigital) [inline app/main/views/questionnaire.py]: does lose some readability by not having the name arguments, what about (skipping obvious named group_id):

```python
this_location = Location(
    group_id, 
    group_instance=0,
    block_id='household-composition')
```

Just for the cases when the parameters being passed in aren't matching names to the function args.

File: app/questionnaire/location.py
Comment (benjefferies) [inline app/questionnaire/location.py]: why self.__dict__ over vars(self)?

File: app/questionnaire/location.py
Comment (collisdigital) [inline app/questionnaire/location.py:12]: don't forget `__hash__`

File: app/questionnaire/navigator.py
Comment (collisdigital) [inline app/questionnaire/navigator.py:55]: maybe this should also be in the Location class as well (if something's opening/closing interstitial?)

File: app/templates/partials/navigation.html
Comment (benjefferies) [inline app/templates/partials/navigation.html]: Would it not be better to expand this when we build the navigation context?

File: app/questionnaire/navigator.py
Comment (collisdigital) [inline app/questionnaire/navigator.py:73]: We can remove this copy now and just modify `this_location` directly as you can use `original_this_location` on line 79 to do the remove, this eliminates the need for `next_location`

File: tests/app/questionnaire/test_navigator.py
Comment (benjefferies) [inline tests/app/questionnaire/test_navigator.py]: Is it not clear enough if we one line the variables, it's a bit noisy otherwise

File: app/questionnaire/location.py
Comment (benjefferies) [inline app/questionnaire/location.py:36]: We should probably have a unit test for this

File: app/questionnaire/navigator.py
Comment (collisdigital) [inline app/questionnaire/navigator.py:91]: What about `Location.from_schema_block(blocks[block_index+1])`

^^ as a static method on `Location` which returns a Location object.

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: app/questionnaire/navigator.py
Comment (collisdigital) [inline app/questionnaire/navigator.py]: Update answer_store to take a Location object?

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py]: Why not?

Comment (benjefferies) [reply app/questionnaire/location.py]: Isn't vars() the pythonic way?

File: tests/app/questionnaire/test_navigator.py
Comment (iwootten) [reply tests/app/questionnaire/test_navigator.py]: Guh, could've asked for these in my other PR, there's 301 of them! See you tomorrow.

File: app/questionnaire/navigator.py
Comment (collisdigital) [inline app/questionnaire/navigator.py]: You could allow `Location.__eq__` to check if `other` is a dict and do a dict comparison so you can do `Location == {...}` ?

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py]: Is it? Never heard of vars and a quick google seems to suggest using __dict__ is the preferred approach.

File: app/questionnaire/location.py
Comment (collisdigital) [inline app/questionnaire/location.py:6]: any value in defaulting these e.g. group_instance = 0 ?

File: app/questionnaire/location.py
Comment (collisdigital) [reply app/questionnaire/location.py]: http://stackoverflow.com/questions/21297203/use-dict-or-vars ?

File: app/questionnaire/navigator.py
Comment (collisdigital) [reply app/questionnaire/navigator.py:91]: I guess that's actually just a `from_dict(...)` method, except the 'block' bit, humph

File: app/questionnaire/location.py
Comment (benjefferies) [reply app/questionnaire/location.py]: From [Fluent Python](https://www.amazon.co.uk/Fluent-Python-Luciano-Ramalho/dp/1491946008)

How Special Methods Are Used
The first thing to know about special methods is that they are meant to be called by the Python interpreter, and not by you. You don’t write `my_object.__len__()`. You write len(my_object) and, if my_object is an instance of a user-defined class, then Python calls the `__len__` instance method you implemented.

File: tests/app/questionnaire/test_navigator.py
Comment (collisdigital) [reply tests/app/questionnaire/test_navigator.py]: The problem is the GUID ids, not the multi-line? Sticking these on one line would be hard to follow I think

File: app/main/views/questionnaire.py
Comment (benjefferies) [inline app/main/views/questionnaire.py]: It's a preference thing but I think it's beneficial to keep Location(...) one line as long as it fits but named params make sense where applicable

File: app/questionnaire/navigator.py
Comment (benjefferies) [reply app/questionnaire/navigator.py:91]: You could have something like
```
def changing(group_id=None, block_id=None, group_instance=None):
    return Answer(group_id or self.group_id, block_id or self.block_id, group_instance or self.group_instance)
```

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py]: Sorry, arbitrary links on the net do nothing for me: http://stackoverflow.com/questions/1227121/compare-object-instances-for-equality-by-their-attributes-in-python

Comment (collisdigital) [reply app/questionnaire/location.py]: https://www.google.co.uk/trends/explore?q=Python%20__dict__,Python%20vars ? 😁

Comment (collisdigital) [reply app/questionnaire/location.py]: I have no opinion on it, just supplying random web content as you suggest.

File: tests/app/questionnaire/test_navigator.py
Comment (iwootten) [reply tests/app/questionnaire/test_navigator.py]: Too late, taken ages and removed them as requested. I tend to agree that we're going to know the order of the args and the real problem is the guids need updating.

File: app/main/views/questionnaire.py
Comment (iwootten) [reply app/main/views/questionnaire.py]: Too late now.

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py:6]: Think it better being explicit. What would .url() do in such a case?

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py:12]: ok

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py]: Yes, as I was yesterday :)

File: app/questionnaire/navigator.py
Comment (iwootten) [reply app/questionnaire/navigator.py:73]: Really? If I make this change it fails all over the place.

File: app/questionnaire/location.py
Comment (collisdigital) [inline app/questionnaire/location.py]: This is automatically defined by Python to be not '__eq__' so no need to add it

File: app/questionnaire/location.py
Comment (jonnyshaw89) [reply app/questionnaire/location.py]: I always favour consistency...... however github won't allow me to post the links

Comment (jonnyshaw89) [reply app/questionnaire/location.py]: __dict__ used 5 times in our code
vars(self) not used in our code

File: app/questionnaire/navigator.py
Comment (collisdigital) [reply app/questionnaire/navigator.py:73]: Not sure then, best leave it!

File: app/questionnaire/questionnaire_manager.py
Comment (collisdigital) [inline app/questionnaire/questionnaire_manager.py]: Same as above comment about `__eq__`

**Comment by collisdigital** [comment]
Sorry Ian, I had started these review comments last week and didn't press the Submit button!

File: app/questionnaire/navigator.py
Comment (iwootten) [reply app/questionnaire/navigator.py]: We'd deliberately be masking a type conversion, which seems wrong...

File: app/main/views/questionnaire.py
Comment (ajmaddaford) [inline app/main/views/questionnaire.py]: Can we change _render_template to take a location parameter? This would look much neater throughout as `_render_template(q_manager.state, this_location, template='questionnaire'`

File: app/questionnaire/location.py
Comment (ajmaddaford) [inline app/questionnaire/location.py:34]: I think we need a better name for these, as we can have blocks of type 'interstitial' in a survey.

File: app/questionnaire/questionnaire_manager.py
Comment (ajmaddaford) [reply app/questionnaire/questionnaire_manager.py]: Maybe we should think about completed blocks being a list of Location objects, and when we serialise/deserialise in QuestionnaireStore we convert it to/from a dict.

File: app/submitter/converter.py
Comment (ajmaddaford) [inline app/submitter/converter.py]: Can this be replaced by `answer_store.filter(location=location)`?

File: app/templating/summary_context.py
Comment (ajmaddaford) [inline app/templating/summary_context.py]: I think using `location` rather than `l` would be clearer here

**Review by ajmaddaford — DISMISSED** [review]

File: tests/app/questionnaire/test_location.py
Comment (benjefferies) [inline tests/app/questionnaire/test_location.py:10]: This setUp shouldn't be needed

File: app/data_model/answer_store.py
Comment (benjefferies) [inline app/data_model/answer_store.py:157]: Shall we remove the :param and :return: documentation if we aren't going to add detail otherwise it's just boilerplate noise

File: app/data_model/answer_store.py
Comment (benjefferies) [inline app/data_model/answer_store.py]: Shall we remove the :param and :return: documentation if we aren't going to add detail otherwise it's just boilerplate noise

File: app/questionnaire/location.py
Comment (benjefferies) [reply app/questionnaire/location.py:36]: Done

File: app/questionnaire/navigator.py
Comment (benjefferies) [inline app/questionnaire/navigator.py]: I think this comment is now stale that we don't keep state of the routing_path

**Review by benjefferies — CHANGES_REQUESTED** [review]
Think there are a few tidy ups we could apply before merging back. It's hard to know what has been actioned/discussed offline etc. It would be good to see a comment to let us know what is happening with those. In general this is a good tidy up

File: tests/app/questionnaire/test_location.py
Comment (iwootten) [reply tests/app/questionnaire/test_location.py:10]: It is, as the app context is necessary to allow the url generation.

File: app/questionnaire/navigator.py
Comment (iwootten) [reply app/questionnaire/navigator.py:91]: Not sure what is being asked here, either solution seems to be more complex than what is in place already.

File: app/questionnaire/location.py
Comment (iwootten) [reply app/questionnaire/location.py:34]: Agreed offline with @ajmaddaford for this to be looked as part of making interstitials proper blocks.

File: app/questionnaire/navigator.py
Comment (collisdigital) [reply app/questionnaire/navigator.py:91]: yeah no clean way to deal with it; one to revisit in future

File: app/questionnaire/navigator.py
Comment (iwootten) [reply app/questionnaire/navigator.py:55]: Agreed offline with @ajmaddaford for this to be looked as part of making interstitials proper blocks.

File: app/questionnaire/navigator.py
Comment (collisdigital) [reply app/questionnaire/navigator.py]: True, although this is implicitly stating that a location `__dict__` matches a block dict structure; i.e. all code using this method of comparison is making that assumption rather than Location object itself doing so. Not sure what's better, leave it to your discretion.

File: app/data_model/answer_store.py
Comment (benjefferies) [inline app/data_model/answer_store.py:164]: 👍

**Review by benjefferies — APPROVED** [review]
Looks good to me!

File: app/data_model/answer_store.py
Comment (collisdigital) [inline app/data_model/answer_store.py:175]: maybe an assert/warning/exception when location  and the individual settings are both set, since its overwriting them here and the caller needs to decide what's being used.

**Review by collisdigital — APPROVED** [review]
Reviewed and Approved

File: app/data_model/answer_store.py
Comment (iwootten) [reply app/data_model/answer_store.py:175]: Added an assert here

**Pull request merged**
